### ISSUE TITLE
Multiple goals simultaneously

### ISSUE THREAD

**Issue by ThatGrzegorzB** [cabaletta/baritone#186]
Possible an option to run multiple goals simultaneously like following a player but mining blocks so depending on which goal it would prioritize, for example if follow player was first it would only mine if it was in range of the player, but if mine was first it would first mine everything it has then go to the player and start mining again

**Comment by ThatGrzegorzB** [comment]
Yes preety much

**Comment by leijurv** [comment]
@ThatGrzegorzB which one? @Thunder33345 gave two possibilities (1: it considers both goals and could go to either one, but it prefers the higher priority one, 2: it goes to the higher priority goal, and only after it gets there does it consider and path to the lower priority goal). Which one was your "Yes preety much" in response to?

**Comment by ThatGrzegorzB** [comment]
Number 2, but it only considers the lower priority goal if the higher priority goal is finished, my idea was using this to follow players and for example mine or find blocks navigate to them and mine another block until the block it was supposed to find is gone

**Comment by SylveonBottle** [comment]
I'd say we should keep #176 separate from #186. Queuing goals to complete in order, and doing things simultaneously seem a bit different.

`#mine log 64 & #goto home` = baritone should walk home, but occasionally mine logs if it gets the chance.  
`#mine log 64 ; #goto home` = baritone should mine logs, then once it's finished walk home.

Though, priority may be depneding on the order.
`#kill pig & #goto home` = Baritone should kill all pigs in sight. If none are found, start walking home until it finds more pigs.
`#goto home & #kill pigs` = Baritone should walk towards home, but if it happens to walk near a pig go ahead and kill it.

You could probably queue it beyond this too, alongside #176:
`#kill pig & #goto home ; #mine diamond_ore` = Baritone should kill pigs that it finds, walking home when there are no targets. Once finally home, start mining diamonds.

Combine with #708 to make a minimum amount of stuff.

`#goto home & #kill pig10 ; #mine diamond_ore 32 ; #goto home` = Baritone should walk home, but kill pigs when it can. If the user has 10 porkchops, ignore pigs and just head home. If the user eats and has less than 10 porkchop (which would happen once the previous pigs are out of render distance), just keep going home until the next time it finds pigs. Once home, mine 32 diamonds. Once mined & collected, head back home once more.

This results in a super long trip back home where baritone is restocking on food by itself, and once home will collect 32 diamonds before heading back home again.

Lengthy post but the possibilities are endless if you manage to add this sort of "baritone goal language".
`#goto portal ; #goal 0 0 ; #invert ; #path & #goto portal ; #goal ; #invert ; #path & #kill pig 32 ; #sethome ; #mine log 32 & mine diamond_ore 32 ; #goto home ; #build house`
1. Go to portal `#goto portal`
2. Go towards a random direction for all eternity basically. `#goal 0 0 ; #invert ; #path`
3. If it finds a portal along the path (eventually), go through it `#path & #goto portal`
4. Go towards a random direction, keep killing pigs ` #goal ; #invert ; #path & #kill pig 32`
5. Once the user ever has 32 pork chops, consider it home `#kill pig 32 ; #sethome`
6. Start mining for logs and diamonds until you have enough `#mine log 32 & mine diamond_ore 32`
7. Go home, construct a house (#112)

And there we go. Baritone goes from being a google maps in minecraft to possibly one of the most advanced minebots that may ever be known, simply from using queued commands.

**Comment by jonnyawsom3** [comment]
if this ever gets implemented, what about also adding the ability to pause and resume certain tasks? (for example, task 1 is "mine some wood" and task 2 is walk away from a point, you say pause 1 and then it'll start walking, say pause 2 and then pause 1 again (or unpause) and it starts getting wood wherever you currently are)

**Comment by SylveonBottle** [comment]
Duplicate of #269, [according to l1ving](https://github.com/cabaletta/baritone/issues/269#issuecomment-751557525)

**Comment by CesiumCs** [comment]
oh my fucking god

**Comment by AirAustin** [comment]
So..
The method of stacking commands doesn't work
"Too many arguments (expected at most 3)"

**Comment by ZacSharp** [comment]
@AirAustin what did you expect from trying to use a feature that does not exist? This is a feature request and not a manual.

**Comment by SylveonBottle** [comment]
I found this issue personally resolved by using [AltoClef](https://github.com/gaucho-matrero/altoclef).

More specifically, [this fork of altoclef](https://github.com/Meloweh/altoclef) which allows running chain commands.

My previous example of `#goto home & #kill pig 10 ; #mine diamond_ore 32 ; #goto home` can be entirely solved by using
```
@roundtrip food 10
@roundtrip give diamond 32
@roundtrip goto X Y Z
@roundtrip start
```

**Comment by AirAustin** [comment]
If it was requested 3 fucking years ago, why isn't it implemented yet

**Comment by SylveonBottle** [comment]
> If it was requested 3 fucking years ago, why isn't it implemented yet

If you want to stack commands, check out [AltoClef](https://github.com/gaucho-matrero/altoclef). Meloweh's fork has `@roundtrip` which is what you want.

**Comment by scorbett123** [comment]
> If it was requested 3 fucking years ago, why isn't it implemented yet

You can implement it if you like. All developers of this project volunteer their time, so it's not always possible to have this kind of issue fixed quickly due to lack of that time.

**Comment by ZacSharp** [comment]
Time is not the issue for simple "run the next command whenever you are idle" scripts.
<details><summary>I spent less time on the code than I spent on a video, which ended up being too large, so here's just the end of it.</summary> The part you see here is<pre>sel clear
sel pos1
sel pos2
sel ex n up 2
sel ex n north 1
sel ex n south 1
sel ex n east 1
sel ex n west 1
sel ca
sel clear
sel pos1
sel pos2
sel sh n down 1
sel ex n east 2
sel ex n west 2
sel ex n north 2
sel ex n south 2
sel pos1
sel pos2
sel sh n up 3
sel ex n east 2
sel ex n west 2
sel ex n north 2
sel ex n south 2
sel pos1
sel pos2
sel sh n east 2
sel ex n up 2
sel ex n north 2
sel ex n south 2
sel pos1
sel pos2
sel sh n west 2
sel ex n up 2
sel ex n north 2
sel ex n south 2
sel pos1
sel pos2
sel sh n south 2
sel ex n up 2
sel ex n east 1
sel ex n west 1
sel pos1
sel pos2
sel sh n north 2
sel sh n west 1
sel ex n up 2
sel pos1
sel pos2
sel sh n north 2
sel sh n east 1
sel ex n up 2
sel pos1
sel pos2
sel sh n north 2
sel sh n up 2
sel ex n up 1
sel fill log
sel clear</pre>
and the cut away part was sylveons last example but without any concurrency (which also means I had to use `#goto` instead of `#invert` and `#path`) 

https://user-images.githubusercontent.com/68165024/148705632-c33e0b80-6df7-4a56-93dc-c2d6c962e616.mp4

</details>

Though before committing it I'd like to know whether scripts with control flow and variables could become a feature as well, because if not I'd simplify a few things like pushing all commands into one queue instead of keeping the macros on a stack each with it's own current command index.
Furthermore I'd really like to get the priority based concurrency, but I'm not sure on how to make it cooperate with the current processes.
